librarymanagement/src/main/scala/sbt/internal/librarymanagement/ivyint/ParallelResolveEngine.scala

```diff
@@ -0,0 +1,102 @@
+
+    val start = System.currentTimeMillis
+    val dependencies0 = report.getDependencies
+    val dependencies = dependencies0
+      .toArray(new Array[IvyNode](dependencies0.size))
+    val artifacts = report.getArtifacts
+      .toArray(new Array[Artifact](report.getArtifacts.size))
+
+    getEventManager.fireIvyEvent(new PrepareDownloadEvent(artifacts))
+
+    // Farm out the dependencies for parallel download
+    val allDownloads = dependencies.par.flatMap { dep =>
+      if (!(dep.isCompletelyEvicted || dep.hasProblem) &&
+          dep.getModuleRevision != null) {
+        ParArray(downloadNodeArtifacts(dep, artifactFilter, options))
+      } else ParArray.empty[DownloadResult]
+    }
+
+    // Force parallel downloads and compute total downloaded size
+    val totalSize = allDownloads.toArray.foldLeft(0L) {
+      case (size, download) =>
+        val dependency = download.dep
+        val moduleConfigurations = dependency.getRootModuleConfigurations
+        moduleConfigurations.foreach { configuration =>
+          val configurationReport = report.getConfigurationReport(configuration)
+
+          // Take into account artifacts required by the given configuration
+          if (dependency.isEvicted(configuration) ||
+              dependency.isBlacklisted(configuration)) {
+            configurationReport.addDependency(dependency)
+          } else configurationReport.addDependency(dependency, download.report)
+        }
+
+        size + download.totalSizeDownloaded
+    }
+
+    report.setDownloadTime(System.currentTimeMillis() - start)
+    report.setDownloadSize(totalSize)
+  }
+
+  /**
+   * Download all the artifacts associated with an ivy node.
+   *
+   * Return the report and the total downloaded size.
+   */
+  private def downloadNodeArtifacts(dependency: IvyNode,
+                                    artifactFilter: Filter,
+                                    options: DownloadOptions): DownloadResult = {
+
+    val resolver = dependency.getModuleRevision.getArtifactResolver
+    val selectedArtifacts = dependency.getSelectedArtifacts(artifactFilter)
+    val downloadReport = resolver.download(selectedArtifacts, options)
+    val artifactReports = downloadReport.getArtifactsReports
+
+    val totalSize = artifactReports.foldLeft(0L) { (size, artifactReport) =>
+      // Check download status and report resolution failures
+      artifactReport.getDownloadStatus match {
+        case DownloadStatus.SUCCESSFUL =>
+          size + artifactReport.getSize
+        case DownloadStatus.FAILED =>
+          val artifact = artifactReport.getArtifact
+          val mergedAttribute = artifact.getExtraAttribute("ivy:merged")
+          if (mergedAttribute != null) {
+            Message.warn(s"\tMissing merged artifact: $artifact, required by $mergedAttribute.")
+          } else {
+            Message.warn(s"\tDetected merged artifact: $artifactReport.")
+            resolver.reportFailure(artifactReport.getArtifact)
```

Comment:
Shouldn't this be outside the if/else block?

Reply:
I guess not, it's identical to the original:

https://github.com/sbt/ivy/blob/b18f59ea3bc914a297bb6f1a4f7fb0ace399e310/src/java/org/apache/ivy/core/resolve/ResolveEngine.java#L392-L399